### DISCUSSION TITLE
Number of Questions regarding CC1101 Register settings

### DISCUSSION THREAD

**Discussion by Crsarmv7l** [jgromes/RadioLib#1020] [Q&A] [answered]
Hi again,

Thank you for the help/responses I received here https://github.com/jgromes/RadioLib/discussions/1016

After experimenting some more with LSatan's library without much luck, I am switching my project to radiolib, especially since it has well developed Sync Serial methods.

I pretty thoroughly went through the Radiolib CC1101 documentation, but there were a few items I did not see. I imagine I will have more questions after this so I plan on containing all future questions I have to this one discussion.

Using Radiolib, how would I adjust the following CC1101 settings (I didn't see a way in the documentation, but I may have missed it)?

- Channel Spacing
- Channel
- Sync mode (EG 16/16 Syncword bits detected, 30/32 Syncword bits detected, Carrier- Sense above threshold, etc) from page 77 of the CC1101 Datasheet in register MDMCFG2
- Enable/disable White Data
- Enable/disable CRC
- Enable/disable DC Filter
- Enable/disable FEC
- Set PQT

Thank you

**Comment by jgromes** [comment]
> Channel Spacing/Channel

RadioLib works with frequencies, not channel numbers, hence these are not implemented.

> Sync mode

See `CC1101::setSyncWord`

> Enable/disable White Data

See `CC1101::setEncoding`

> Enable/disable CRC

See `CC1101::setCrcFiltering`

> Enable/disable DC Filter

Not sure what you menat by that. Assuming you want to have DC-free data encoding (such as Manchester encoding), then see `CC1101::setEncoding`

> Enable/disable FEC

This is not implemented.

> Set PQT

See `CC1101::setPreambleLength`

**Comment by Crsarmv7l** [comment]
I see most everything is defined in cc1101.h anything to stop me from using
say SPIsetRegValue(RADIOLIB_CC1101_REG_MDMCFG2,
RADIOLIB_CC1101_DEM_DCFILT_OFF)

**Reply by jgromes** [reply]
Yes, the fact that `SPIsetRegValue` is deliberately a private method, you cannot use outside of RadioLib.

You can however enable god mode, e.g. by adding `#define RADIOLIB_GODMODE (1)` to `BuildOptUser.h`, which will allow acces to all private/protected members, including direct acess to SPI.

**Reply by Crsarmv7l** [reply]
Cool thanks. I will have to put together a simple ino to see if I can get the basics working, then see how I can work it all together

**Reply by Crsarmv7l** [reply]
@jgromes  Here I am here again with another question. Thank you for your patience, and I do appreciate your help. Sorry meant to respond to you but didn't, apologies for the tag.

On all of your examples for the CC1101 you tend to have something similar to this:

// CC1101 connections
// cs = 5
// GDO0 = 11
// Reset = not needed
// GDO2 = 13 (optional)
CC1101 radio = new Module(5, 11, RADIOLIB_NC, 13);

I notice in this discussion regarding direct mode: https://github.com/jgromes/RadioLib/discussions/312

Clock is on GDO0 and Data on GDO2, which is fine with me, while GDO1 is not in the setup. That raises the question below though:

In the direct sx1278 here https://github.com/jgromes/RadioLib/blob/master/examples/SX127x/SX127x_Receive_Direct/SX127x_Receive_Direct.ino

I noticed the pin setup is DIO0 and DIO1, and special setup for the DIO2 pin which is later read by radio.readBit(pin);

My Questions are these: Is GDO0 clock and GDO2 data for the CC1101? If so, since GDO0 and GDO2 are already defined in the CC1101 radio = new Module how is this handled with radio.readBit?

**Reply by jgromes** [reply]
> Is GDO0 clock and GDO2 data for the CC1101?

Yes it is. You don't have to take my word for it, it's in the source:

https://github.com/jgromes/RadioLib/blob/cfc425970cbcb2355c0d8c02d6d589228c0d144e/src/modules/CC1101/CC1101.cpp#L988-L991

> since GDO0 and GDO2 are already defined in the CC1101 radio = new Module how is this handled with radio.readBit?

Functions of CC1101 GDOx pins do not map onto DIOx pins of SX127x. The GPIO pins taken by `Module` are typically used for signalling packet being sent or received. To have at least some commonality between different radios, `Module` constructor provides two possible GPIO/IRQ pins for all RadioLib modules. On SX127x, it just so happens that DIO2 is needed for direct mode (and this functionality cannot be performed by DIO0 on DIO1, that's why it's initialized outside the library. This is also the reason why `readBit` takes the pin number. 

TL;DR: Even in the case of CC1101, you have to call `readBit(GDO2)` to sample the pin.

**Comment by Crsarmv7l** [comment]
> Yes, the fact that `SPIsetRegValue` is deliberately a private method, you cannot use outside of RadioLib.
> 
> You can however enable god mode, e.g. by adding `#define RADIOLIB_GODMODE (1)` to `BuildOptUser.h`, which will allow acces to all private/protected members, including direct acess to SPI.

Hmm with regard to my above questions, perhaps the answer is to not define GD02 in the CC1101 radio = new Module(5, 11, RADIOLIB_NC, 13); leaving it free to define as DIO2 is in the sx1278 example?

**Comment by Crsarmv7l** [comment]
Makes sense thank you.

Last question. I assume I will have to use setGdo0Action in order to only
sample/readbit when the clock is high?

**Reply by jgromes** [reply, marked as answer]
You should use `setDirectAction` so that the CC1101 driver sets the correct pin (GDO0) and interrupt direction (rising edge).

**Reply by Crsarmv7l** [reply]
Thank you so much for all your help.
